### PR TITLE
remove unused pub fns

### PR DESCRIPTION
**Pull request by klensy**
This removes some unused `pub fn`; also fixes few obsoleted fn names or added fixmes with reminders to update them.

### CODE REVIEW

File: compiler/rustc_middle/src/mir/query.rs [hunk 1]
```diff
@@ -42,64 +42,6 @@ pub enum UnsafetyViolationDetails {
     CallToFunctionWith,
 }
 
-impl UnsafetyViolationDetails {
-    pub fn description_and_note(&self) -> (&'static str, &'static str) {
```
Comment (klensy) [inline compiler/rustc_middle/src/mir/query.rs:46]: Big nice function but no usage, sus.

File: compiler/rustc_const_eval/src/interpret/projection.rs [hunk 2]
```diff
@@ -255,13 +255,15 @@ where
         self.project_index(base, index)
     }
 
+    // FIXME: fix comment `operand_index` - no such fn
     /// Iterates over all fields of an array. Much more efficient than doing the
     /// same by repeatedly calling `operand_index`.
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/projection.rs:260]:
```suggestion
    /// same by repeatedly calling `project_index`.
```

File: compiler/rustc_const_eval/src/interpret/projection.rs [hunk 3]
```diff
@@ -255,13 +255,15 @@ where
...
     pub fn project_array_fields<'a, P: Projectable<'tcx, M::Provenance>>(
         &self,
         base: &'a P,
     ) -> InterpResult<'tcx, ArrayIterator<'tcx, 'a, M::Provenance, P>> {
         let abi::FieldsShape::Array { stride, .. } = base.layout().fields else {
+            // FIXME: fix comment `operand_array_fields` - no such fn
             span_bug!(self.cur_span(), "operand_array_fields: expected an array layout");
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/projection.rs:267]:
```suggestion
            span_bug!(self.cur_span(), "project_array_fields: expected an array layout");
```

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 4]
```diff
@@ -998,6 +998,7 @@ impl<'a, 'tcx> ProofTreeVisitor<'tcx> for AmbiguityCausesVisitor<'a> {
 
         let Goal { param_env, predicate } = goal.goal();
 
+        // FIXME: no `replace_bound_vars_with_placeholders` fn exist
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:1001]: This is https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.instantiate_binder_with_placeholders

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 5]
```diff
@@ -329,6 +329,7 @@ fn equate_impl_headers<'tcx>(
                 impl1.self_ty,
                 impl2.self_ty,
             ),
+            // FIXME: mk_eq_impl_headers fn not exist
```
Comment (davidtwco) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:332]: I presume this is just now `equate_impl_headers`, the current function.

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 6]
```diff
@@ -615,6 +615,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return false;
         }
         let pin_did = self.tcx.lang_items().pin_type();
+        // FIXME: replace mk_box with?
```
Comment (davidtwco) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:618]: `Ty::new_box` as per below?

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after the two comments are addressed

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs (hunk 6 above)
Comment (klensy) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:618]: There still no `unwrap` here, should it be `This guards the new_box below.` or remove comment completely?

Comment (klensy) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs]: added review commit.

**Comment by klensy** [comment]
Rebased, should i squash?

**Pull request merged**